contracts/strategies/morpho/MetaMorphoStrategy.sol

```diff
@@ -287,4 +312,85 @@ contract MetaMorphoStrategy is ERC4626Strategy {
         _rewardData.marketId
       ).supplyRewardTokensPerYear / _oneReward;
   }
+
+  /// @notice calculate how much of vault `_add` amount will be added to this market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _supplyQueueLen supply queue length
+  /// @param _add amount of liquidity to add
+  function _calcMarketAdd(
+    IMMVault _mmVault,
+    bytes32 _targetMarketId,
+    uint256 _supplyQueueLen,
+    uint256 _add
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    // loop throuh supplyQueue, starting from the first market, and see how much will
+    // be deposited in target market
+    for (uint256 i = 0; i < _supplyQueueLen; i++) {
+      bytes32 _currMarketId = _mmVault.supplyQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      // get max depositable amount for this market
+      (uint184 _marketCap,,) = _mmVault.config(_targetMarketId);
+      uint256 _maxDeposit = uint256(_marketCap) - _market.totalSupplyAssets;
+      // If this is the target market, return the current _add value, eventually
+      // reduced to the max depositable amount
+      if (_currMarketId == _targetMarketId) {
+        if (_add > _maxDeposit) {
+          _add = _maxDeposit;
+        }
+        break;
+      }
+      // If this is not the target market, check if we can deposit all the _add amount
+      // in this market, otherwise continue the loop and subtract the max depositable
+      if (_add > _maxDeposit) {
+        _add -= _maxDeposit;
+      } else {
+        _add = 0;
+        break;
+      }
+    }
+
+    return _add;
+  }
+
+  /// @notice calculate how much of vault `_sub` amount will be removed from target market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _withdrawQueueLen withdraw queue length
+  /// @param _sub liquidity to remove
+  function _calcMarketSub(
+    IMMVault _mmVault, 
+    bytes32 _targetMarketId,
+    uint256 _withdrawQueueLen,
+    uint256 _sub
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    // loop throuh withdrawQueue, and see how much will be redeemed in target market
+    for (uint256 i = 0; i < _withdrawQueueLen; i++) {
+      bytes32 _currMarketId = _mmVault.withdrawQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      // get available liquidity for this market
+      uint256 _availableLiquidity = _market.totalSupplyAssets - _market.totalBorrowAssets;
```

Comment:
note that we use [zeroFloorSub](https://github.com/morpho-org/metamorpho/blob/32d5ae71af4d8560cfde6fcd5b41d798c9c06406/src/MetaMorpho.sol#L841) in case of roundings (very unlikely)

Reply:
maybe I can add a check for `_market.totalBorrowAssets > _market.totalSupplyAssets`?

Reply:
added the mentioned check